### PR TITLE
Custom `mean!` for `AbstractReducedField`

### LINKED ISSUE #1767: Output of `AveragedField` is sometimes wrong on the GPU
It seems that with Julia 1.6 and recent versions of CUDA.jl, the output of averaged fields is wrong. By inspection it seems that the output is being written to disk prematurely somehow as some grid points contain the sum instead of the mean.

@xkykai and I have tried to reproduce in CUDA.jl but with no luck so far.

I've tried adding a `CUDA.@sync blocking=true` to https://github.com/CliMA/Oceananigans.jl/blob/383173d11a0c96182a4349fc1e33755207bf0886/src/Fields/averaged_field.jl#L62 but that didn't help.

Adding a `sleep(0.01)` afterwards helped a bit (reduced the number of occurences of this issue). It's not a useful hack but it may suggest that a race condition is involved?

X-Ref: This is a minimal reproduction of https://github.com/CliMA/LESbrary.jl/issues/118
X-Ref: https://github.com/JuliaGPU/CUDA.jl/issues/929 (Attempt at reproducing but turned out to be a REPL issue)

# Minimal working example

```julia
using Oceananigans

grid = RegularRectilinearGrid(size=(256, 256, 128), extent=(1, 1, 1))
model = IncompressibleModel(architecture=GPU(), grid=grid)
simulation = Simulation(model, Δt=1, stop_time=100)

set!(model, u=1, v=1, T=1)

U = AveragedField(model.velocities.u, dims=(1, 2))
V = AveragedField(model.velocities.v, dims=(1, 2))
T = AveragedField(model.tracers.T, dims=(1, 2))

simulation.output_writers[:jld2] =
    JLD2OutputWriter(model, (; U, V, T),
          prefix = "stats",
        schedule = TimeInterval(10),
           force = true)

simulation.output_writers[:nc] =
    NetCDFOutputWriter(model, (; U, V, T),
        filepath = "stats.nc",
        schedule = TimeInterval(10),
            mode = "c")

run!(simulation)
```

## Looking at the NetCDF output

```julia
using NCDatasets
ds = NCDataset("stats.nc")
ds["T"][:]
```

```
128×11 Matrix{Float32}:                                                                                                                                                                                                                                                                 
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      0.0      0.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 ⋮                                ⋮                                            ⋮
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
```

## Looking at the JLD2 output

```julia
using JLD2
ds = jldopen("stats.jld2")
T_jld2 = cat([ds["timeseries/T/$iter"][:] for iter in keys(ds["timeseries/t"])]..., dims=2)
```

```
128×11 Matrix{Float32}:                                                                                                                                                                                                                                                                 
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      0.0      0.0      1.0      1.0      1.0                                                                                                                                                                                                      
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 1.0  1.0  1.0      1.0      1.0  1.0      1.0      1.0      1.0      1.0      1.0
 ⋮                                ⋮                                            ⋮
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
 1.0  1.0  1.0  65536.0  65536.0  1.0  65536.0  65536.0  65536.0  65536.0  65536.0
```

### PR DESCRIPTION
**Pull request by glwagner**
Computes normalization via broadcasting over `parent(R::AbstractReducedField)` rather than `R` itself. May resolve #1767.

### CODE REVIEW

**Review by ali-ramadhan — APPROVED** [review]
Ah interesting. Does this work because broadcasting over GPU subarray views is a little rough around the edges?

Also, is it worth adding the MWE from https://github.com/CliMA/Oceananigans.jl/issues/1767#issuecomment-868793917 as a test?

File: src/Fields/mapreduce_abstract_fields.jl [hunk 1]
```diff
@@ -14,3 +15,10 @@ for function_name in (:sum, :prod, :maximum, :minimum, :all, :any)
         Base.$(function_name!)(r::AbstractReducedField, a::AbstractArray; kwargs...) = Base.$(function_name!)(identity, interior(r), a; kwargs...)
     end
 end
+
+function Statistics.mean!(R::AbstractReducedField, A::AbstractArray)
```
Comment (ali-ramadhan) [inline src/Fields/mapreduce_abstract_fields.jl:19]:
```suggestion
# For the reason why this dispatch is needed, see: https://github.com/CliMA/Oceananigans.jl/issues/1767
function Statistics.mean!(R::AbstractReducedField, A::AbstractArray)
```

**Comment by glwagner** [comment]
> Ah interesting. Does this work because broadcasting over GPU subarray views is a little rough around the edges?
> 
> Also, is it worth adding the MWE from [#1767 (comment)](https://github.com/CliMA/Oceananigans.jl/issues/1767#issuecomment-868793917) as a test?

Certainly _something_ is rough... I think it's "broadcasting over ReducedField" that's the issue here. Prior to this PR, `mean!` used a broadcast over `R::AbstractReducedField` to compute the normalization. For some reason this has data synchronicity issues on the GPU (I'm at a loss to explain why). This PR just changes that broadcast operation to compute over all members of `R`. The halo regions of `R` (presumably) aren't touched during `sum!`, so doing some extra normalization in the halos doesn't really matter, I guess... (if we want halos to be right we should probably fill them after executing `mean!` in `compute!`, or something).

But the _reason_ why this change fixes the issue isn't necessarily obvious. Maybe there's a bad interaction between `KernelAbstractions` (which gets used for the broadcast) and `sum!`? Not sure.

**Comment by glwagner** [comment]
We do have tests for `AveragedField` right? Do we need a more expensive / larger test (like the one proposed in #1767) to catch this bug?

**Comment by ali-ramadhan** [comment]
> We do have tests for `AveragedField` right?

Yes but they didn't catch #1767.

> Do we need a more expensive / larger test (like the one proposed in #1767) to catch this bug?

I think so. #1767 seems like an important bug to catch since it's a pretty serious bug and could crop up again in the future.

**Comment by ali-ramadhan** [comment]
I'm not sure. I guess whatever reproduces #1767. It seems that it only occurs sometimes so maybe we just need to make an existing test larger and run it multiple times, like 10 times?

**Comment by glwagner** [comment]
I'll open a test PR from master and see if I can get a test to fail!

**Comment by navidcy** [comment]
i vote for a test to catch this also. 

And yes, patch release please!

**Review by navidcy — APPROVED** [review]
I also vote for a test to catch this also.

And yes, patch release please!

**Comment by glwagner** [comment]
The tricky thing is that this PR is supposed to fix the issue. So we need a second PR with a test that fails, and then to cherry-pick that test into this PR (after confirming manually that the test fails for the "right" reason). Anybody is welcome to open this other PR.

**Comment by tomchor** [comment]
Any reason why this PR hasn't been merged? This seems like a pretty important bug to fix so I'd vote for a new bug fix version with this asap.

**Comment by glwagner** [comment]
We're trying to design a test that catches the bug this PR fixes. But since this PR _fixes_ the bug, we have to do that in another PR. I've been working on that at #1784. I'd be ok merging this PR without that test and adding the test later, since all the existing tests pass for this PR. What do you think @ali-ramadhan ?

**Comment by glwagner** [comment]
The difficulties in getting a test to fail on #1784 could mean that the issue this PR resolves is machine dependent. But I'm not sure.

**Review by ali-ramadhan — APPROVED** [review]
Yeah we can continue trying to find a test in #1784. Otherwise agree with @tomchor that merging soon and tagging a new release would be good.

**Comment by tomchor** [comment]
(Anxiously waiting for this to merge so that I can update my scripts :))

**Comment by glwagner** [comment]
> (Anxiously waiting for this to merge so that I can update my scripts :))

I'll merge but just to be sure, the scripts can be updated provided you depend on `Oceananigans#glw/be-meaner` in your environment.

**Comment by glwagner** [comment]
🤞 hope GPU works since we don't have GPU CI right now...

**Comment by tomchor** [comment]
> > (Anxiously waiting for this to merge so that I can update my scripts :))
> 
> I'll merge but just to be sure, the scripts can be updated provided you depend on `Oceananigans#glw/be-meaner` in your environment.

I'm aware. I just was waiting for this because master already has some other bug fixes that I want, and it hadn't been merged with this branch.

> crossed_fingers hope GPU works since we don't have GPU CI right now...

Hmm, that's a good point. Now that you merged master into this branch it might be worth waiting and I'll just use this branch.

**Comment by glwagner** [comment]
If it's urgent, I suggest creating a new branch that merges master and this one and depending on that.

**Comment by tomchor** [comment]
> I suggest creating a new branch that merges master and whatever other branch has an important feature

That's a good point. Hadn't thought of that option

**Pull request merged**